graphdatascience/procedure_surface/api/similarity/knn_endpoints.py

```diff
@@ -0,0 +1,434 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any
+
+from pandas import DataFrame
+
+from graphdatascience.procedure_surface.api.base_result import BaseResult
+from graphdatascience.procedure_surface.api.catalog.graph_api import GraphV2
+from graphdatascience.procedure_surface.api.estimation_result import EstimationResult
+
+
+class KnnMutateResult(BaseResult):
+    """Represents the result of running K-Nearest Neighbors in mutate mode."""
+
+    pre_processing_millis: int
+    compute_millis: int
+    mutate_millis: int
+    post_processing_millis: int
+    nodes_compared: int
+    relationships_written: int
+    similarity_distribution: dict[str, Any]
+    did_converge: bool
+    ran_iterations: int
+    node_pairs_considered: int
+    configuration: dict[str, Any]
+
+
+class KnnStatsResult(BaseResult):
+    """Represents the result of running K-Nearest Neighbors in stats mode."""
+
+    pre_processing_millis: int
+    compute_millis: int
+    post_processing_millis: int
+    nodes_compared: int
+    similarity_pairs: int
+    similarity_distribution: dict[str, Any]
+    did_converge: bool
+    ran_iterations: int
+    node_pairs_considered: int
+    configuration: dict[str, Any]
+
+
+class KnnWriteResult(BaseResult):
+    """Represents the result of running K-Nearest Neighbors in write mode."""
+
+    pre_processing_millis: int
+    compute_millis: int
+    write_millis: int
+    post_processing_millis: int
+    nodes_compared: int
+    relationships_written: int
+    did_converge: bool
+    ran_iterations: int
+    node_pairs_considered: int
+    similarity_distribution: dict[str, Any]
+    configuration: dict[str, Any]
+
+
+class KnnEndpoints(ABC):
+    @abstractmethod
+    def mutate(
+        self,
+        G: GraphV2,
+        mutate_relationship_type: str,
+        mutate_property: str,
+        node_properties: str | list[str] | dict[str, str],
+        top_k: int | None = None,
+        similarity_cutoff: float | None = None,
+        delta_threshold: float | None = None,
+        max_iterations: int | None = None,
+        sample_rate: float | None = None,
+        perturbation_rate: float | None = None,
+        random_joins: int | None = None,
+        random_seed: int | None = None,
+        initial_sampler: Any | None = None,
+        relationship_types: list[str] | None = None,
+        node_labels: list[str] | None = None,
+        sudo: bool | None = None,
+        log_progress: bool = True,
+        username: str | None = None,
+        concurrency: Any | None = None,
```

Comment:
concurrency, jobId should have a type, just like in the other endpoints. 
initialSampler is probably a string